raitools/raitools/_managers/counterfactual_manager.py

```diff
@@ -2,33 +2,225 @@
+                    _CounterfactualConstants.OPPOSITE:
+                raise UserConfigValidationException(
+                    'The desired_class attribute should not be \'{0}\''
+                    ' It should be the class value for multiclass'
+                    ' classification scenario.'.format(
+                        _CounterfactualConstants.OPPOSITE))
+
+        if self._task_type == ModelTask.REGRESSION:
+            if new_counterfactual_config.desired_range is None:
+                raise UserConfigValidationException(
+                    'The desired_range should not be None'
+                    ' for regression scenarios.')
+
+        duplicate_counterfactual_config_found = False
+        for counterfactual_config in self._counterfactual_config_list:
+            if counterfactual_config == new_counterfactual_config:
+                duplicate_counterfactual_config_found = True
+                break
+
+        if duplicate_counterfactual_config_found:
+            raise DuplicateCounterfactualConfigException(
+                'Duplicate counterfactual configuration detected.')
+        else:
+            self._counterfactual_config_list.append(new_counterfactual_config)
 
     def add(self,
-            continuous_features=None,
-            outcome_name=None,
-            total_CFs=None,
+            continuous_features,
+            total_CFs,
+            method=_CounterfactualConstants.RANDOM,
+            desired_class=None,
             desired_range=None,
-            desired_class="opposite",
             permitted_range=None,
-            features_to_vary=None,
-            method=None):
-        raise NotImplementedError(
-            "Add not implemented for CounterfactualManager")
+            features_to_vary=None):
+        """Add a counterfactual generation configuration to be computed later.
+
+        :param continuous_features: List of names of continuous features.
+                                    The remaining features are categorical
+                                    features.
+        :type continuous_features: List
+        :param total_CFs: Total number of counterfactuals required.
+        :type total_CFs: int
+        :param desired_class: Desired counterfactual class. For binary
+                              classification, this needs to be set as
+                              "opposite".
+        :type desired_class: string or int
+        :param desired_range: For regression problems.
+                              Contains the outcome range
+                              to generate counterfactuals in.
+        :type desired_range: List
+        :param permitted_range: Dictionary with feature names as keys and
+                                permitted range in list as values.
+                                Defaults to the range inferred from training
+                                data.
+        :type permitted_range: Dict
```

Comment:
lowercase dict/list for types to follow conventions?

Reply:
Done